src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -0,0 +1,540 @@
+            .map(|(_, shard_id)| shard_id)
+            .collect();
+        old_shard_ids.extend(
+            migration_shards_to_finalize
+                .iter()
+                .map(|(_, shard_id)| shard_id),
+        );
+        txn.insert_unfinalized_shards(old_shard_ids).map_err(|e| {
+            Error::new(ErrorKind::Internal(format!(
+                "builtin migration failed: {e}"
+            )))
+        })?;
+        txn.insert_collection_metadata(global_id_shards)
+            .map_err(|e| {
+                Error::new(ErrorKind::Internal(format!(
+                    "builtin migration failed: {e}"
+                )))
+            })?;
+    }
+
+    let updates = txn.get_and_commit_op_updates();
+    let builtin_table_updates = state.apply_updates_for_bootstrap(updates).await;
+
+    let cleanup_action = async move {
+        if !read_only {
+            let upper = fetch_upper(&mut write_handle).await;
+            let updates: Vec<_> = migration_shards_to_finalize
+                .into_iter()
+                .map(|(table_key, shard_id)| ((table_key, shard_id), upper, -1))
+                .collect();
+            // Ignore any errors, these shards will get cleaned up in the next upgrade.
+            let res =
+                write_to_migration_shard(updates, upper, &mut write_handle, &mut since_handle)
+                    .await;
+            if let Err(e) = res {
+                error!("Unable to remove old entries from migration shard: {e:?}");
+            }
+        }
+    }
+    .boxed();
+
+    Ok(BuiltinItemMigrationResult {
+        builtin_table_updates,
+        storage_collections_to_drop: BTreeSet::new(),
+        cleanup_action,
+    })
+}
+
+async fn fetch_upper(
+    write_handle: &mut WriteHandle<TableKey, ShardId, Timestamp, Diff>,
+) -> Timestamp {
+    write_handle
+        .fetch_recent_upper()
+        .await
+        .as_option()
+        .cloned()
+        .expect("we use a totally ordered time and never finalize the shard")
+}
+
+async fn write_to_migration_shard(
+    updates: Vec<((TableKey, ShardId), Timestamp, Diff)>,
+    upper: Timestamp,
+    write_handle: &mut WriteHandle<TableKey, ShardId, Timestamp, Diff>,
+    since_handle: &mut SinceHandle<TableKey, ShardId, Timestamp, Diff, i64>,
+) -> Result<(), Error> {
+    let next_upper = upper.step_forward();
+    // Lag the shard's upper by 1 to keep it readable.
+    let downgrade_to = Antichain::from_elem(next_upper.saturating_sub(1));
+    let next_upper = Antichain::from_elem(next_upper);
+
+    if let Err(_) = write_handle
+        .compare_and_append(updates, Antichain::from_elem(upper), next_upper)
+        .await
+        .expect("invalid usage")
+    {
+        return Err(Error::new(ErrorKind::Durable(DurableCatalogError::Fence(
+            "Catalog fenced during builtin table migrations".to_string(),
+        ))));
+    }
+
+    // The since handle gives us the ability to fence out other writes using an opaque token.
```

Comment:
```suggestion
    // The since handle gives us the ability to fence out other downgraders using an opaque token.
```

? Writer feels too write-y :man_shrugging: